### PR TITLE
Deprecate the `-fbasic-block-sections=labels` option.

### PR DESCRIPTION
**Pull request by rlavaee**
This feature is supported via the newer option `-fbasic-block-address-map`. Using the old option still works by delegating to the newer option, while a warning is printed to show deprecation.

### CODE REVIEW

**Review by mtrofin — APPROVED** [review]
nice!

**Review by red1bluelost — APPROVED** [review]
LGTM

File: clang/docs/UsersManual.rst [hunk 1]
```diff
@@ -2369,14 +2369,16 @@ are listed below.
      $ cd $P/bar && clang -c -funique-internal-linkage-names name_conflict.c
      $ cd $P && clang foo/name_conflict.o && bar/name_conflict.o
 
-.. option:: -fbasic-block-sections=[labels, all, list=<arg>, none]
+.. option:: -f[no]-basic-block-address-map:
+  Emits a ``SHT_LLVM_BB_ADDR_MAP`` which includes address offsets for each
```
Comment (MaskRay) [inline clang/docs/UsersManual.rst:2373]: `SHT_LLVM_BB_ADDR_MAP section`

File: lld/ELF/LTO.cpp [hunk 2]
```diff
@@ -64,14 +64,16 @@ static lto::Config createConfig() {
...
-  // Allowed values for --lto-basic-block-sections are "all", "labels",
+  // Allowed values for --lto-basic-block-sections are "all",
   // "<file name specifying basic block ids>", or none.  This is the equivalent
   // of -fbasic-block-sections= flag in clang.
   if (!config->ltoBasicBlockSections.empty()) {
     if (config->ltoBasicBlockSections == "all") {
       c.Options.BBSections = BasicBlockSection::All;
     } else if (config->ltoBasicBlockSections == "labels") {
-      c.Options.BBSections = BasicBlockSection::Labels;
+      c.Options.BBAddrMap = true;
```
Comment (MaskRay) [inline lld/ELF/LTO.cpp:74]: suggest that you create a separate change for lld/ELF.
If this is deprecated, we need a test for the warning message.

Please check other warnings for the style. We don't add the full stop.

**Review by MaskRay — APPROVED** [review]
(Thanks for the patch. I am on vacation and may be slow/inconvenient to respond.)

**Comment by jh7370** [comment]
Just chiming in that I happened to spot the pre-merge check failure looks possibly related.

**Comment by rlavaee** [comment]
> Just chiming in that I happened to spot the pre-merge check failure looks possibly related.

@jh7370 It's weird I can't reproduce this on my machine.

**Comment by rlavaee** [comment]
> > Just chiming in that I happened to spot the pre-merge check failure looks possibly related.
> 
> @jh7370 It's weird I can't reproduce this on my machine.

Fixed. It was an assertion failure.

**Comment by zeroomega** [comment]
Hi

This change likely broke the clang build:

```
[4323/4602](116) Building CXX object tools/lld/ELF/CMakeFiles/lldELF.dir/LTO.cpp.o
 FAILED: tools/lld/ELF/CMakeFiles/lldELF.dir/LTO.cpp.o 
 /b/s/w/ir/x/w/rc/cxx-rbeirddhris/reclient-cxx-wrapper.sh /b/s/w/ir/x/w/cipd/bin/clang++ --sysroot=/b/s/w/ir/x/w/cipd/linux -DGTEST_HAS_RTTI=0 -DLLD_VENDOR=\"Fuchsia\" -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/b/s/w/ir/x/w/llvm_build/tools/lld/ELF -I/b/s/w/ir/x/w/llvm-llvm-project/lld/ELF -I/b/s/w/ir/x/w/llvm-llvm-project/lld/include -I/b/s/w/ir/x/w/llvm_build/tools/lld/include -I/b/s/w/ir/x/w/rc/tensorflow-venv/store/python_venv-nacb1phkpgklqiq46g3ivnt518/contents/lib/python3.8/site-packages/tensorflow/include -I/b/s/w/ir/x/w/llvm_build/include -I/b/s/w/ir/x/w/llvm-llvm-project/llvm/include -isystem /b/s/w/ir/x/w/install-zlib-x86_64-linux-gnu/include -isystem /b/s/w/ir/x/w/install-zstd-x86_64-linux-gnu/include -stdlib=libc++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -ffat-lto-objects -ffile-prefix-map=/b/s/w/ir/x/w/llvm_build=../llvm-llvm-project -ffile-prefix-map=/b/s/w/ir/x/w/llvm-llvm-project/= -no-canonical-prefixes -O3 -DNDEBUG -std=c++17 -fvisibility=default  -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti -MD -MT tools/lld/ELF/CMakeFiles/lldELF.dir/LTO.cpp.o -MF tools/lld/ELF/CMakeFiles/lldELF.dir/LTO.cpp.o.d -o tools/lld/ELF/CMakeFiles/lldELF.dir/LTO.cpp.o -c /b/s/w/ir/x/w/llvm-llvm-project/lld/ELF/LTO.cpp
 /b/s/w/ir/x/w/llvm-llvm-project/lld/ELF/LTO.cpp:74:49: error: no member named 'Labels' in 'llvm::BasicBlockSection'
    74 |       c.Options.BBSections = BasicBlockSection::Labels;
       |                              ~~~~~~~~~~~~~~~~~~~^
 1 error generated.
```

See build task: https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-base-linux-x64/b8735797591984119265/overview

Could you revert your change please?

**Pull request merged**